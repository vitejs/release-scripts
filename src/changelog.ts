import fs from "node:fs";
import path from "node:path";
import { ConventionalChangelog, type Preset } from "conventional-changelog";
import createPreset, {
  DEFAULT_COMMIT_TYPES,
  // @ts-expect-error no types
} from "conventional-changelog-conventionalcommits";
import type { generateChangelog as def } from "./types.d.ts";

export const generateChangelog: typeof def = async ({
  getPkgDir,
  tagPrefix,
}) => {
  const preset: Preset = await createPreset({
    types: DEFAULT_COMMIT_TYPES.map((t: any) => ({ ...t, hidden: false })),
  });
  preset.writer ??= {};
  preset.writer.headerPartial = `
## {{#if isPatch~}} <small> {{~/if~}}
{{#if @root.linkCompare~}}
[{{version}}](
{{~#if @root.repository~}}
  {{~#if @root.host}}
    {{~@root.host}}/
  {{~/if}}
  {{~#if @root.owner}}
    {{~@root.owner}}/
  {{~/if}}
  {{~@root.repository}}
{{~else}}
  {{~@root.repoUrl}}
{{~/if~}}
/compare/{{previousTag}}...{{currentTag}})
{{~else}}
{{~version}}
{{~/if}}
{{~#if title}} "{{title}}"
{{~/if}}
{{~#if date}} ({{date}})
{{~/if}}
{{~#if isPatch~}} </small> {{~/if}}
`.trim();
  preset.writer.mainTemplate =
    `
{{> header}}
{{#if noteGroups}}
{{#each noteGroups}}

### âš  {{title}}

{{#each notes}}
* {{#if commit.scope}}**{{commit.scope}}:** {{/if}}{{commit.subject}} {{#if commit.hash}}([{{commit.shortHash}}](https://github.com/{{@root.owner}}/{{@root.repository}}/commit/{{commit.hash}})){{/if}}
{{/each}}
{{/each}}
{{/if}}
{{#each commitGroups}}

{{#if title}}
### {{title}}

{{/if}}
{{#each commits}}
{{> commit root=@root}}
{{/each}}
{{/each}}`.trim() + "\n";

  const pkgDir = getPkgDir();

  const generator = new ConventionalChangelog(pkgDir)
    .readPackage()
    .config(preset)
    .options({ releaseCount: 1 });
  if (tagPrefix) {
    generator.tags({ prefix: tagPrefix });
  }

  const originalChangelog = fs.existsSync(path.join(pkgDir, "CHANGELOG.md"))
    ? fs.readFileSync(path.join(pkgDir, "CHANGELOG.md"), "utf-8")
    : "";

  const writeStream = fs.createWriteStream(path.join(pkgDir, "CHANGELOG.md"));
  for await (const chunk of generator.write()) {
    writeStream.write(chunk);
  }
  writeStream.write(originalChangelog);
};
