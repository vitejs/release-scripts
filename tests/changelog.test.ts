import fs from "node:fs/promises";
import { it, expect, onTestFinished } from "vitest";
import { createFixture, type FileTree } from "fs-fixture";
import { execa } from "execa";
import { generateChangelog } from "../src/changelog.ts";

async function createProjectFixture(source?: FileTree) {
  const fixture = await createFixture({
    "package.json": JSON.stringify({
      name: "test-project",
      version: "1.0.0",
      private: true,
    }),
    ...source,
  });
  onTestFinished(() => fixture.rm());

  const initialCwd = process.cwd();
  process.chdir(fixture.path);
  onTestFinished(() => process.chdir(initialCwd));

  await execa("git", ["init"]);
  await execa("git", [
    "remote",
    "add",
    "origin",
    "https://github.com/vitejs/test.git",
  ]);

  return fixture;
}

async function gitCommit(message: string) {
  // Write random text to file to allow conventional-changelog to detect commit
  await fs.writeFile("dummy.txt", Math.random().toString(36).substring(2, 15));
  await execa("git", ["add", "."]);
  await execa("git", ["commit", "-m", message]);
}

async function updatePackageJsonVersion(version: string) {
  const pkgPath = "./package.json";
  const pkg = JSON.parse(await fs.readFile(pkgPath, "utf8"));
  pkg.version = version;
  await fs.writeFile(pkgPath, JSON.stringify(pkg));
}

async function initChangelog() {
  await gitCommit("chore: initial commit");
  await generateChangelogForRelease();
}

async function generateChangelogForRelease() {
  await generateChangelog({ getPkgDir: () => ".", tagPrefix: "" });

  // Tag the version so conventional-changelog tracks this as the last release
  // and won't track this commit for the next release.
  const version = JSON.parse(
    await fs.readFile("./package.json", "utf8"),
  ).version;
  const tag = `v${version}`;
  await execa("git", ["tag", "-a", "-m", tag, tag]);
}

async function readChangelog() {
  const changelog = await fs.readFile("./CHANGELOG.md", "utf8");
  return (
    changelog
      // Normalize date
      .replace(/\d{4}-\d{2}-\d{2}/g, "yyyy-mm-dd")
      // Normalize short commit hashes
      .replace(/\[[a-z0-9]{7}\]/g, `[${"x".repeat(7)}]`)
      // Normalize full commit hashes
      .replace(/\/[a-z0-9]{40}\)/g, `/${"x".repeat(40)})`)
  );
}

it("generates a new changelog for empty project", async () => {
  await createProjectFixture();
  await gitCommit("chore: initial commit");
  await generateChangelogForRelease();
  expect(await readChangelog()).toMatchSnapshot();
});

it("generates a changelog with commits", async () => {
  await createProjectFixture();
  await initChangelog();

  await gitCommit("fix: fix a bug (#1)");
  await updatePackageJsonVersion("1.0.1");
  await generateChangelogForRelease();
  expect(await readChangelog()).toMatchSnapshot();

  await gitCommit("feat: add new feature");
  await updatePackageJsonVersion("1.1.0");
  await generateChangelogForRelease();
  expect(await readChangelog()).toMatchSnapshot();
});

it("generates a changelog with breaking changes", async () => {
  await createProjectFixture();
  await initChangelog();
  await gitCommit("feat!: introduce breaking change");
  await gitCommit("fix: fix a bug (#1)");
  await updatePackageJsonVersion("2.0.0");
  await generateChangelogForRelease();
  expect(await readChangelog()).toMatchSnapshot();
});
