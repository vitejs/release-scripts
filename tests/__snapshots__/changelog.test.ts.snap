// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a changelog with breaking changes 1`] = `
"## [2.0.0](https://github.com/vitejs/test/compare/v1.0.0...v2.0.0) (yyyy-mm-dd)
### âš  BREAKING CHANGES

* introduce breaking change ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))

### Features

* introduce breaking change ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))

### Bug Fixes

* fix a bug ([#1](https://github.com/vitejs/test/issues/1)) ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
## 1.0.0 (yyyy-mm-dd)
### Miscellaneous Chores

* initial commit ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
"
`;

exports[`generates a changelog with commits 1`] = `
"## <small>[1.0.1](https://github.com/vitejs/test/compare/v1.0.0...v1.0.1) (yyyy-mm-dd)</small>
### Bug Fixes

* fix a bug ([#1](https://github.com/vitejs/test/issues/1)) ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
## 1.0.0 (yyyy-mm-dd)
### Miscellaneous Chores

* initial commit ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
"
`;

exports[`generates a changelog with commits 2`] = `
"## [1.1.0](https://github.com/vitejs/test/compare/v1.0.1...v1.1.0) (yyyy-mm-dd)
### Features

* add new feature ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
## <small>[1.0.1](https://github.com/vitejs/test/compare/v1.0.0...v1.0.1) (yyyy-mm-dd)</small>
### Bug Fixes

* fix a bug ([#1](https://github.com/vitejs/test/issues/1)) ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
## 1.0.0 (yyyy-mm-dd)
### Miscellaneous Chores

* initial commit ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
"
`;

exports[`generates a new changelog for empty project 1`] = `
"## 1.0.0 (yyyy-mm-dd)
### Miscellaneous Chores

* initial commit ([xxxxxxx](https://github.com/vitejs/test/commit/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx))
"
`;
